File: epub-extraction-tool/extract.js
const { EPub } = require('epub2');
const fs = require('fs-extra');
const path = require('path');

async function extractEpubInfo(epub_path, output_info_dir, output_covers_dir) {
    try {
        // Use createAsync for modern epub2 library
        const epub = await EPub.createAsync(epub_path);
        
        const basename = path.basename(epub_path, path.extname(epub_path));
        console.log(`Processing: ${basename}`);

        // 1. Prepare metadata
        const bookInfo = {
            title: epub.metadata.title,
            authors: Array.isArray(epub.metadata.creator) ? epub.metadata.creator : [epub.metadata.creator],
            language: epub.metadata.language,
            description: epub.metadata.description || "",
            publishDate: epub.metadata.date || "",
            publisher: epub.metadata.publisher || "",
            tags: epub.metadata.subject || [],
        };

        // 2. Save book info to JSON
        await fs.ensureDir(output_info_dir);
        const jsonPath = path.join(output_info_dir, `${basename}.json`);
        await fs.writeJson(jsonPath, bookInfo, { spaces: 4 });
        console.log(`  - Book info saved to: ${jsonPath}`);

        // 3. Save book cover
        await fs.ensureDir(output_covers_dir);
        if (epub.metadata.cover) {
            const data = await new Promise((resolve, reject) => {
                epub.getImage(epub.metadata.cover, (err, data) => {
                    if (err) return reject(err);
                    resolve(data);
                });
            });

            // Try to find the original extension in manifest
            const coverManifestItem = epub.manifest[epub.metadata.cover];
            const ext = coverManifestItem && coverManifestItem['media-type'] ? (coverManifestItem['media-type'].split('/')[1] || 'jpg') : 'jpg';
            const coverPath = path.join(output_covers_dir, `${basename}.${ext}`);
            
            await fs.writeFile(coverPath, data);
            console.log(`  - Cover saved to: ${coverPath}`);
        } else {
            console.warn(`  - Warning: No cover image found in ${basename}.`);
        }

    } catch (err) {
        console.error(`  - Error processing ${epub_path}:`, err);
    }
}

async function main() {
    const args = process.argv.slice(2);
    if (args.length === 0) {
        console.error("Usage: node extract.js <input_path> [--info-dir=extracted_info] [--covers-dir=extracted_covers]");
        process.exit(1);
    }

    const input_path = args[0];
    const infoDir = args.find(arg => arg.startsWith('--info-dir='))?.split('=')[1] || 'extracted_info';
    const coversDir = args.find(arg => arg.startsWith('--covers-dir='))?.split('=')[1] || 'extracted_covers';

    if (fs.statSync(input_path).isFile() && input_path.endsWith('.epub')) {
        await extractEpubInfo(input_path, infoDir, coversDir);
    } else if (fs.statSync(input_path).isDirectory()) {
        const files = await fs.readdir(input_path);
        const epubs = files.filter(f => f.endsWith('.epub'));
        if (epubs.length === 0) {
            console.log(`No EPUB files found in ${input_path}.`);
            return;
        }
        console.log(`Found ${epubs.length} EPUB files. Starting extraction...`);
        for (const epub of epubs) {
            await extractEpubInfo(path.join(input_path, epub), infoDir, coversDir);
        }
    } else {
        console.error(`Error: ${input_path} is not a valid file or directory.`);
    }
}

main();

File: epub-extraction-tool/extract.py
import os
import json
import argparse
from ebookmeta import get_metadata
from pathlib import Path

def extract_epub_info(epub_path, output_info_dir, output_covers_dir):
    try:
        epub_path = Path(epub_path)
        if not epub_path.is_file():
            print(f"Error: {epub_path} is not a valid file.")
            return

        print(f"Processing: {epub_path.name}")
        metadata = get_metadata(str(epub_path))

        # 1. Prepare metadata
        authors = []
        if hasattr(metadata, 'authors') and metadata.authors:
            authors = [a.author for a in metadata.authors] if hasattr(metadata.authors, '__iter__') else [metadata.author_list]
        else:
            authors = [metadata.author_list] if metadata.author_list else []

        book_info = {
            "title": metadata.title,
            "authors": authors,
            "language": metadata.lang,
            "description": metadata.description or "",
            "publish_date": metadata.publish_date or "",
            "publisher": metadata.publisher or "",
            "tags": str(metadata.tag_list).split(';') if metadata.tag_list else [],
            "series": metadata.series or "",
            "series_index": metadata.series_index or "",
        }

        # 2. Save book info to JSON
        os.makedirs(output_info_dir, exist_ok=True)
        json_filename = f"{epub_path.stem}.json"
        json_path = os.path.join(output_info_dir, json_filename)
        with open(json_path, 'w', encoding='utf-8') as f:
            json.dump(book_info, f, indent=4, ensure_ascii=False)
        print(f"  - Book info saved to: {json_path}")

        # 3. Save book cover
        os.makedirs(output_covers_dir, exist_ok=True)
        if metadata.cover_image_data:
            # Determine extension
            ext = 'jpg' if 'jpeg' in (metadata.cover_media_type or 'jpeg').lower() else 'png'
            cover_filename = f"{epub_path.stem}.{ext}"
            cover_path = os.path.join(output_covers_dir, cover_filename)
            
            with open(cover_path, 'wb') as f:
                f.write(metadata.cover_image_data)
            print(f"  - Cover saved to: {cover_path}")
        else:
            print("  - Warning: No cover image found in this EPUB.")

    except Exception as e:
        print(f"  - Error processing {epub_path.name}: {e}")

def main():
    parser = argparse.ArgumentParser(description="Extract EPUB metadata and covers into separate folders.")
    parser.add_argument("input", help="Path to a single EPUB file or a directory containing EPUB files.")
    parser.add_argument("--info-dir", default="extracted_info", help="Directory where book info JSONs will be saved.")
    parser.add_argument("--covers-dir", default="extracted_covers", help="Directory where book covers will be saved.")

    args = parser.parse_args()

    input_path = Path(args.input)

    if input_path.is_file():
        extract_epub_info(input_path, args.info_dir, args.covers_dir)
    elif input_path.is_dir():
        epub_files = list(input_path.glob("*.epub"))
        if not epub_files:
            print(f"No EPUB files found in directory: {input_path}")
            return
        
        print(f"Found {len(epub_files)} EPUB files. Starting extraction...")
        for epub in epub_files:
            extract_epub_info(epub, args.info_dir, args.covers_dir)
    else:
        print(f"Error: {input_path} is not a valid file or directory.")

if __name__ == "__main__":
    main()
